File: QuintaSemana/bandidos.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_set>
#include <algorithm>

using namespace std;

vector<vector<int>> gerar_grafo_invertido(const vector<int>& grafo) {
    int n = grafo.size() - 1;
    vector<vector<int>> grafo_inv(n + 1);
    for (int no = 2; no <= n; no++) {
        int superior = grafo[no];
        if (superior != 0)
            grafo_inv[superior].push_back(no);
    }
    return grafo_inv;
}

vector<int> maior_altura(int raiz, const vector<vector<int>>& grafo_inv) {
    int n = grafo_inv.size() - 1;
    vector<int> altura(n + 1, 0);
    queue<pair<int, int>> q;
    q.push({raiz, 1});
    while (!q.empty()){
        auto [atual, h] = q.front();
        q.pop();
        altura[atual] = h;
        for (int child : grafo_inv[atual]) {
            q.push({child, h + 1});
        }
    }
    vector<int> nodes;
    for (int i = 1; i <= n; i++){
        nodes.push_back(i);
    }
    sort(nodes.begin(), nodes.end(), [&altura](int a, int b) {
        return altura[a] > altura[b];
    });
    return nodes;
}

vector<int> pegar_folhas(const vector<int>& grafo) {
    int n = grafo.size() - 1;
    unordered_set<int> pais;
    for (int i = 1; i <= n; i++){
        pais.insert(grafo[i]);
    }
    vector<int> folhas;
    for (int i = 1; i <= n; i++){
        if (pais.find(i) == pais.end())
            folhas.push_back(i);
    }
    return folhas;
}

int pegar_altura(int no, const vector<int>& grafo, vector<bool>& visitados) {
    int c = 0;
    while (no != 1) {
        if (!visitados[no]) {
            c++;
            visitados[no] = true;
        }
        no = grafo[no];
        if (no == 1 && !visitados[1]) {
            c++;
            visitados[1] = true;
        }
    }
    return c;
}

int prender(const vector<int>& grafo, int nSelecionados) {
    vector<int> folhas = pegar_folhas(grafo);
    unordered_set<int> folhasSet(folhas.begin(), folhas.end());
    vector<bool> visitados(grafo.size(), false);
    
    vector<vector<int>> grafo_inv = gerar_grafo_invertido(grafo);
    vector<int> sorted_nodes = maior_altura(1, grafo_inv);
    
    vector<int> alturas;
    for (int node : sorted_nodes) {
        if (folhasSet.count(node))
            alturas.push_back(pegar_altura(node, grafo, visitados));
    }
    
    sort(alturas.begin(), alturas.end(), greater<int>());
    int resultado = 0;
    for (int i = 0; i < nSelecionados && i < (int)alturas.size(); i++){
        resultado += alturas[i];
    }
    return resultado;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int qtd_membros, previsoes;
    cin >> qtd_membros >> previsoes;
    
    vector<int> grafo(qtd_membros + 1);
    grafo[1] = 1;
    for (int i = 2; i <= qtd_membros; i++){
        cin >> grafo[i];
    }
    
    cout << prender(grafo, previsoes) << "\n";
    return 0;
}

File: 05_QuintaSemana/bandidos.py
from collections import defaultdict
from heapq import nlargest


def prender(grafo, n):
    folhas = pegar_folhas(grafo)
    visitados = [0] * (len(grafo) + 1)
    maiores_alturas = maior_altura(1, gerar_grafo_invertido(grafo))
    alturas = [pegar_altura(a,grafo,visitados) for a in maiores_alturas if a in folhas]
    return sum(nlargest(n,alturas))


def pegar_altura(no, grafo, visitados):
    c = 0
    while no != 1:
        if visitados[no] == 0:
            c += 1
            visitados[no] = 1
        no = grafo[no]
        if no == 1 and visitados[1] == 0:
            c += 1
            visitados[1] = 1
    return c


def gerar_grafo_invertido(grafo):
    grafo_invertido = defaultdict(list)

    for no, superior in grafo.items():
        if superior != 0:
            grafo_invertido[superior].append(no)
    return grafo_invertido


def maior_altura(no, grafo):
    pilha = [(no,1)]
    visitados = set()
    alturas = {}

    while pilha:

        no, altura = pilha.pop()
        if no not in visitados:
            visitados.add(no)
            alturas[no] = altura
            for vizinho in grafo[no]:
                if vizinho not in visitados:
                    pilha.append((vizinho, altura+1))
    return dict(sorted(alturas.items(), key=lambda x: x[1], reverse=True))



def pegar_folhas(grafo):
    filhos = set(grafo.values())
    folhas = [k for k in grafo if k not in filhos]
    return folhas


def main():
    qtd_membros, previsoes = map(int, input().split())
    superiores = input().split()
    grafo = {1: 1}
    for i, superior in enumerate(superiores, start=2):
        grafo[i] = int(superior)
    print(prender(grafo, previsoes))


if __name__ == '__main__':
    main()
